### PR TITLE
Make CSG shape dirty after changing Snap

### LINKED ISSUE #93240: Changing CSG node Snap doesn't regenerate mesh
### Tested versions

- https://github.com/godotengine/godot/pull/91748 (df33cbff3b4a68941debc1c9b2c8fe9b3ba7d97d)
- Godot v4.2.2.stable

### System information

Godot v4.2.2.stable - Windows 10.0.22631 - Vulkan (Forward+)

### Issue description

I found this on a build of https://github.com/godotengine/godot/pull/91748, but it applies to 4.2.2 as well and the code is the same on tip.

I tried running TextMesh through CSGMesh3D and subtracting it from a box ([Text.tscn](https://gist.github.com/31/c3c4bace42d4ca550ae6badf0f2b7cf9)) as a way to generate situations where CSG meshing might have problems without making a bunch of nodes or making a big mesh of my own. I spotted some bumpy triangles:

image

Changing the Box's Snap from 0.001 to 0.0001 and reloading the scene fixes it:

image

But: should I really need to reload the scene? Or should `set_snap` include `_make_dirty();` like the other methods do?

https://github.com/godotengine/godot/blob/71699e08c9df78b7203fa4ef9cede28e995d6ace/modules/csg/csg_shape.cpp#L143-L145

https://github.com/godotengine/godot/blob/71699e08c9df78b7203fa4ef9cede28e995d6ace/modules/csg/csg_shape.cpp#L628-L632

I can imagine someone seeing this problem, not knowing to reload the scene (or jiggle a node) and thinking that Snap is broken! With the improvements to CSG in https://github.com/godotengine/godot/pull/91748, this might interfere with users attempting to try out the feature and make them think it's less capable than it is.

---

On 4.2.2, 0.001 snap looks worse, and the mesh has holes (a known issue across many scenarios):

image

Changing to 0.0001 snap doesn't look quite as good as https://github.com/godotengine/godot/pull/91748, but it is an improvement in this part of the mesh as there are no visible holes:

image

(There are plenty of holes elsewhere in the mesh, but I think that's to be expected.)

### Steps to reproduce

1. Load https://gist.github.com/31/c3c4bace42d4ca550ae6badf0f2b7cf9, or make a CSG scene with problems that need a lower snap value.
2. Change snap.
3. See that the generated mesh doesn't change.

### Minimal reproduction project (MRP)

Text.tscn: https://gist.github.com/31/c3c4bace42d4ca550ae6badf0f2b7cf9

### PR DESCRIPTION
**Pull request by 31**
* Fixes https://github.com/godotengine/godot/issues/93240

### CODE REVIEW

**Review by smix8 — APPROVED** [review]

**Comment by Calinou** [comment]
So that's why I thought changing Snap didn't have any effect…

Thanks for fixing this after so long :slightly_smiling_face:

**Review by Calinou — COMMENTED** [review]
Tested locally, it works as expected.

https://github.com/godotengine/godot/assets/180032/86b925d6-a597-450c-b10d-edcc4b388d65

While we're at it, I'd change CSGShape3D's `snap` documentation to mention that only the top-level CSG node' Snap property is taken into account. Child nodes' value is ignored for snapping vertices of the final (combined) CSG shape.

File: modules/csg/csg_shape.cpp [hunk 1]
```diff
@@ -142,6 +142,7 @@ bool CSGShape3D::is_root_shape() const {
 
 void CSGShape3D::set_snap(float p_snap) {
 	snap = p_snap;
```
Comment (Calinou) [inline modules/csg/csg_shape.cpp:144]: Prevent `_make_dirty()` from being called if the value hasn't actually changed:

```suggestion
void CSGShape3D::set_snap(float p_snap) {
	if (p_snap == snap) {
		return;
	}
	
	snap = p_snap;
```

**Review by fire — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**